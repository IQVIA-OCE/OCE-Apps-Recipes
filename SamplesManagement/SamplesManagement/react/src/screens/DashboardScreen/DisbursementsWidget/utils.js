import moment from 'moment';
const DATE_FORMAT = 'YYYY-MM';
import Color from 'color';

const base_colors = [
  Color('#5899DA'),
  Color('#E8743B'),
  Color('#19A979'),
  Color('#ED4A7B'),
  Color('#945ECF'),
  Color('#13A4B4'),
  Color('#525DF4'),
  Color('#BF399E'),
  Color('#6C8893'),
  Color('#EE6868'),
  Color('#2F6497'),
];

export const generateColors = total_colors => {
  const colors = [];
  const gradient_step = (base_colors.length - 1) / total_colors;

  let gradient = 0;

  colors.push(base_colors[0].hex());

  for (let i = 1; i < total_colors - 1; i++) {
    gradient += gradient_step;

    const from_index = Math.floor(gradient);
    let to_index = Math.ceil(gradient);

    if (from_index === to_index) {
      to_index = from_index + 1;
    }

    const from_color = base_colors[from_index];
    const to_color = base_colors[to_index];

    const mix_position = gradient % 1;

    colors.push(from_color.mix(to_color, mix_position).hex());
  }

  colors.push(base_colors[base_colors.length - 1].hex());

  return colors;
};

const generateDumpChartData = period => {
  const dumpData = [];
  for (let i = period; i >= 0; i -= 1) {
    dumpData.push({
      x: moment()
        .subtract(i, 'months')
        .format(DATE_FORMAT),
      y: 0,
    });
  }

  return dumpData;
};
export const normalizeDisbursements = period => data => {
  const normalized = data.reduce(
    (acc, disbursement) => {
      if (acc.allIds.indexOf(disbursement.Id) === -1) {
        acc.allIds.push(disbursement.Id);
      }

      if (!acc.byId[disbursement.Id]) {
        acc.byId[disbursement.Id] = {
          label: disbursement.productName ? disbursement.productName : '',
          data: generateDumpChartData(period),
        };
      }

      acc.byId[disbursement.Id].data.map(el => {
        const label = moment()
          .month(disbursement.month - 1)
          .year(disbursement.year)
          .format(DATE_FORMAT);

        if (moment(el.x).isSame(label, 'month')) {
          el.y = disbursement.totalQuantity;
        }

        el.tooltip = `${disbursement.productName}: ${el.y}`;
        el.x = moment(el.x).toDate();
        return el;
      });

      return acc;
    },
    { allIds: [], byId: {} }
  );

  normalized.colors = generateColors(normalized.allIds.length)

  normalized.allIds.forEach((id, i) => {
    normalized.byId[id].color = normalized.colors[i];
  })

  return normalized;
};
export  const handleFilterData = (id, filtered, disbursements) => {
  const index = filtered.ids.indexOf(id);
  const colors = [];
  let ids = [];
  if (index !== -1) {
    ids = filtered.ids.filter(el => {
      if (el !== id) {
        colors.push(disbursements.data.byId[el].color);
        return true;
      }
      return false;
    });
  } else {
    disbursements.data.allIds.forEach(el => {
      if (el === id || filtered.ids.includes(el)) {
        ids.push(el);
        colors.push(disbursements.data.byId[el].color);
      }
    });
  }

  return { colors, ids }
};