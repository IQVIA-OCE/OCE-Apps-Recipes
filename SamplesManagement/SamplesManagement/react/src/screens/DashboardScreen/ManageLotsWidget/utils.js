export const normalizeLots = keys => data => {
  const normalize = (item) => {
    const clean = {};
    Object.keys(item).map(key => {
      let value = item[key]
      if (item[key] && typeof item[key] === 'object') {
        value = normalize(item[key])
      }
      clean[keys[key] || key] = value;
    });
    return clean;
  }

  return data.map(item => {
    if (!item || typeof item !== 'object') return;
    return normalize(item);
  });
};
