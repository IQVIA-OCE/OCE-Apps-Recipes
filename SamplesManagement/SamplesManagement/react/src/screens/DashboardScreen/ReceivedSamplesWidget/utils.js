import { NAMESPACE } from '../../../constants/constants';

export const mapReceivedSamples = transactionsData => {
  return transactionsData.map(sampleTransaction => {
    const detailsCount = sampleTransaction[`${NAMESPACE}SampleTransactionItems__r`]
      ? sampleTransaction[`${NAMESPACE}SampleTransactionItems__r`].totalSize
      : null;

    return {
      id: sampleTransaction.Id,
      conditionOfPackage: sampleTransaction[`${NAMESPACE}ConditionOfPackage__c`],
      fromSalesRepId: sampleTransaction[`${NAMESPACE}FromSalesRep__c`],
      fromSalesRepName: sampleTransaction[`${NAMESPACE}FromSalesRep__c`]
        ? sampleTransaction[`${NAMESPACE}FromSalesRep__r`].Name
        : null,
      lastModifiedDate: sampleTransaction.LastModifiedDate,
      receivedDate: sampleTransaction[`${NAMESPACE}ReceivedDate__c`],
      recordTypeName: sampleTransaction.RecordType.Name,
      recordTypeDevName: sampleTransaction.RecordType.DeveloperName,
      shipmentDate: sampleTransaction[`${NAMESPACE}ShipmentDate__c`],
      status: sampleTransaction[`${NAMESPACE}Status__c`],
      toSalesRepId: sampleTransaction[`${NAMESPACE}ToSalesRep__c`],
      toSalesRepName: sampleTransaction[`${NAMESPACE}ToSalesRep__r`]
        ? sampleTransaction[`${NAMESPACE}ToSalesRep__r`].Name
        : null,
      transactionDateTime: sampleTransaction[`${NAMESPACE}TransactionDateTime__c`],
      transactionRepId: sampleTransaction[`${NAMESPACE}TransactionRep__c`],
      transactionRepName: sampleTransaction[`${NAMESPACE}TransactionRep__r`]
        ? sampleTransaction[`${NAMESPACE}TransactionRep__r`].Name
        : null,
      accountId: sampleTransaction[`${NAMESPACE}Call__r`]
        ? sampleTransaction[`${NAMESPACE}Call__r`][`${NAMESPACE}Account__c`]
        : null,
      recordTypeId: sampleTransaction.RecordTypeId,
      accountName:
        sampleTransaction[`${NAMESPACE}Call__r`] &&
        sampleTransaction[`${NAMESPACE}Call__r`][`${NAMESPACE}Account__r`]
          ? sampleTransaction[`${NAMESPACE}Call__r`][`${NAMESPACE}Account__r`].Name
          : null,
      address: sampleTransaction[`${NAMESPACE}FullAddress__c`],
      detailsCount,
    };
  });
};
