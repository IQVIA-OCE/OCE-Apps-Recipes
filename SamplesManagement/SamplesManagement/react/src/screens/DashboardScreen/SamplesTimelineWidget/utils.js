export const mapTransactions = transactionsData => {
  return transactionsData.map(sampleTransaction => {
    const detailsCount = sampleTransaction.OCE__SampleTransactionItems__r
      ? sampleTransaction.OCE__SampleTransactionItems__r.totalSize
      : null;

    return {
      id: sampleTransaction.Id,
      conditionOfPackage: sampleTransaction.OCE__ConditionOfPackage__c,
      fromSalesRepId: sampleTransaction.OCE__FromSalesRep__c,
      fromSalesRepName: sampleTransaction.OCE__FromSalesRep__c
        ? sampleTransaction.OCE__FromSalesRep__r.Name
        : null,
      lastModifiedDate: sampleTransaction.LastModifiedDate,
      receivedDate: sampleTransaction.OCE__ReceivedDate__c,
      recordTypeName: sampleTransaction.RecordType.Name,
      recordTypeDevName: sampleTransaction.RecordType.DeveloperName,
      shipmentDate: sampleTransaction.OCE__ShipmentDate__c,
      status: sampleTransaction.OCE__Status__c,
      toSalesRepId: sampleTransaction.OCE__ToSalesRep__c,
      toSalesRepName: sampleTransaction.OCE__ToSalesRep__r
        ? sampleTransaction.OCE__ToSalesRep__r.Name
        : null,
      transactionDateTime: sampleTransaction.OCE__TransactionDateTime__c,
      transactionRepId: sampleTransaction.OCE__TransactionRep__c,
      transactionRepName: sampleTransaction.OCE__TransactionRep__r
        ? sampleTransaction.OCE__TransactionRep__r.Name
        : null,
      accountId: sampleTransaction.OCE__Call__r
        ? sampleTransaction.OCE__Call__r.OCE__Account__c
        : null,
      recordTypeId: sampleTransaction.RecordTypeId,
      accountName:
        sampleTransaction.OCE__Call__r &&
        sampleTransaction.OCE__Call__r.OCE__Account__r
          ? sampleTransaction.OCE__Call__r.OCE__Account__r.Name
          : null,
      address: sampleTransaction.OCE__FullAddress__c,
      detailsCount,
    };
  });
};

export const mapOrders = ordersData => {
  return ordersData.map(sampleOrder => {
    const detailsCount =
      sampleOrder.OCE__SampleOrderDetails__r != null
        ? sampleOrder.OCE__SampleOrderDetails__r.totalSize
        : null;

    return {
      id: sampleOrder.Id,
      isUrgent: sampleOrder.OCE__IsUrgent__c,
      status: sampleOrder.OCE__Status__c,
      lastModifiedDate: sampleOrder.LastModifiedDate,
      recordTypeName: 'Order',
      recordTypeDevName: 'Order',
      detailsCount,
    };
  });
};
