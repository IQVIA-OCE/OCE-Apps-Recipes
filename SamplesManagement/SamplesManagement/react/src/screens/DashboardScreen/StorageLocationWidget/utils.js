export const normalizeLocation = records =>
  records.length ? records[0]['OCE__FullAddress__c'] : '';
