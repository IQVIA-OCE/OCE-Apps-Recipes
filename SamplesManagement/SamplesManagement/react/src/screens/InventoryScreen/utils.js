import moment from 'moment';
import { INVENTORY_STATUS } from './constants';

const AOS_RECORDTYPE = 'AcknowledgementOfShipment';
const RETURN_RECORDTYPE = 'Return';
const TRANSFEROUT_RECORDTYPE = 'TransferOut';
const TRANSFERIN_RECORDTYPE = 'TransferIn';
const ADJUSTMENT_RECORDTYPE = 'Adjustment';

export const normalizeProductsList = (
  products,
  lastSubmittedInventoriesDetails,
  inventoriesDetails,
  transactions,
  status,
  exist
) => {
  const selectedProducts = [];

  const normalizedProducts = products.map(el => {
    let obj = {
      deleted: false,
      locked: false,
      hasQuantityError: false,
      discrepancyReason: '',
    };

    obj.lotNumberId = el.Id;
    obj.sampleProductId = el.OCE__Product__c;
    obj.detailLabel = el.Name;
    obj.lotNumber = el.Name;
    obj.label = el.OCE__Product__r ? el.OCE__Product__r.Name : '';
    obj.quantityIn = 0;
    obj.quantityOut = 0;
    obj.openingBalance = 0;
    obj.systemCount = 0;
    obj.physicalQuantity = null;

    lastSubmittedInventoriesDetails &&
      lastSubmittedInventoriesDetails.length &&
      lastSubmittedInventoriesDetails.forEach(inventory => {
        if (inventory.OCE__Lot__c === el.Id) {
          if (!obj.locked) {
            obj.locked = true;
            selectedProducts.push(obj);
          }

          if (
            inventory.OCE__Status__c === INVENTORY_STATUS.submitted ||
            inventory.OCE__Status__c === INVENTORY_STATUS.saved ||
            (inventory.OCE__Status__c === INVENTORY_STATUS.inProgress && exist)
          ) {
            obj.openingBalance = inventory.OCE__PhysicalCount__c;
            obj.systemCount = inventory.OCE__PhysicalCount__c;
          }
        }
      });

    inventoriesDetails &&
      inventoriesDetails.length &&
      inventoriesDetails.forEach(inventory => {
        if (inventory.OCE__Lot__c === el.Id) {
          if (!obj.locked) {
            if (status === INVENTORY_STATUS.inProgress && exist) {
              obj.selected = true;
            } else {
              obj.locked = true;
            }
            selectedProducts.push(obj);
          }
          obj.id = inventory.Id;
          obj.physicalQuantity =
            inventory.OCE__PhysicalCount__c &&
            inventory.OCE__PhysicalCount__c.toString();
          obj.discrepancyReason = inventory.OCE__DiscrepancyReason__c;
        }
      });

    if (status !== INVENTORY_STATUS.submitted && transactions.length) {
      transactions.forEach(transaction => {
        if (transaction.OCE__LotNumber__c === el.Id) {
          if (!obj.locked) {
            // product should be pushed into array if it neither locked nor selected
            if (!obj.selected) {
              selectedProducts.push(obj);
            }
            obj.locked = true;
          }

          obj = getQuantity(transaction, obj);
        }
      });
    }

    return obj;
  });

  return [normalizedProducts, selectedProducts];
};

export const normalizeProductsHistoryList = products =>
  products.map(product => {
    return {
      recordType: product.OCE__SampleTransaction__r.RecordType,
      quantity: product.OCE__Quantity__c,
      transactionDateTime: moment
        .utc(product.OCE__SampleTransaction__r.OCE__TransactionDateTime__c)
        .format('MMM D, YYYY hh:mm A'),
      productName: product.OCE__Product__r ? product.OCE__Product__r.Name : '',
      productId: product.OCE__Product__c,
      lotId: product.OCE__LotNumber__c,
      lotNumber: product.OCE__LotNumber__r.Name,
    };
  });

export const getQuantity = (transaction, obj) => {
  switch (transaction.OCE__SampleTransaction__r.RecordType.DeveloperName) {
    case AOS_RECORDTYPE:
    case TRANSFERIN_RECORDTYPE:
      obj.quantityIn += transaction.OCE__Quantity__c;
      break;

    case ADJUSTMENT_RECORDTYPE:
      if (transaction.OCE__Quantity__c > 0) {
        obj.quantityIn += transaction.OCE__Quantity__c;
      } else {
        obj.quantityOut += transaction.OCE__Quantity__c * -1;
      }
      break;
    case RETURN_RECORDTYPE:
    case TRANSFEROUT_RECORDTYPE:
      obj.quantityOut += transaction.OCE__Quantity__c;
      break;
  }

  obj.systemCount = obj.openingBalance + obj.quantityIn - obj.quantityOut;

  return obj;
};

export const checkProductsInvalid = products =>
  products.find(el => el.systemCount !== parseInt(el.physicalQuantity));
