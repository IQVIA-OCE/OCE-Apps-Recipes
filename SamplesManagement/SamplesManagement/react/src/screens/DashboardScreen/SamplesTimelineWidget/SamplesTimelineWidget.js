import React, { useCallback } from 'react';
import { View } from 'react-native';
import {
  Card,
  IconMenuButton,
  Tabs,
  ActivityIndicator,
  useTheme,
} from '@oce-apps/apollo-react-native';
import TimelineComponent from './timeline/Timeline';
import moment from 'moment';
import ViewAll from '../../../components/ViewAll/ViewAll';
import {
  fetchOrdersData,
  fetchOrdersListId,
  fetchTransactionsData,
  fetchTransactionsListId,
} from '../../../api/SamplesTimeline';
import { mapOrders, mapTransactions } from './utils';
import { useNavigation, useRoute, useFocusEffect } from '@react-navigation/native';
import { NAMESPACE } from '../../../constants/constants';

function FetchData({ route, navigation, onUpdate }) {
  useFocusEffect(
    useCallback(() => {
      if (route.params && route.params.refreshSamplesTimelineWidget) {
        onUpdate();
      }
      return () => {
        navigation.setParams({
          refreshSamplesTimelineWidget: false,
        });
      };
    }, [route.params?.refreshSamplesTimelineWidget])
  );
  return null;
}

class SamplesTimelineWidget extends React.Component {
  state = {
    transactionsList: [],
    ordersList: [],
    allItemsList: [],
    transactionsListId: null,
    isLoading: false,
    syncDateLabel: '',
  };

  focusListener = null;

  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.getTimeLineData(6);
  }

  getTimeLineData = async limit => {
    try {
      this.setState({ isLoading: true });
      const transactionsData = await fetchTransactionsData(limit);
      const ordersData = await fetchOrdersData(limit);
      const transactionsList = mapTransactions(transactionsData[0]);
      const transactionsListId = await fetchTransactionsListId();
      const ordersListId = await fetchOrdersListId();
      const ordersList = mapOrders(ordersData[0]);
      const allItemsList = [...transactionsList, ...ordersList].sort(
        (a, b) =>
          moment(b.lastModifiedDate).valueOf() -
          moment(a.lastModifiedDate).valueOf()
      );

      this.setState({
        transactionsList,
        transactionsListId: transactionsListId[0][0].Id,
        ordersListId: ordersListId[0][0].Id,
        ordersList,
        allItemsList: allItemsList.slice(0, 6),
        isLoading: false,
        syncDateLabel: moment().format('dddd MMM D, YYYY hh:mm A'),
      });
    } catch (error) {
      console.log(error);
      this.setState({
        isLoading: false,
      });
    }
  };

  render() {
    const {
      transactionsList,
      ordersList,
      allItemsList,
      syncDateLabel,
      transactionsListId,
      ordersListId,
      isLoading,
    } = this.state;

    const {route, navigation, theme} = this.props;

    const menuItems = [
      {
        text: 'Transaction',
        onPress: () => this.props.navigation.navigate('RecordTypeSelector'),
      },
      {
        text: 'Order',
        onPress: () => this.props.navigation.navigate('SampleOrder'),
      },
    ];

    const subtitle = syncDateLabel ? `As of ${syncDateLabel}` : '';

    return (
      <>
        <FetchData navigation={navigation} route={route} onUpdate={() => this.getTimeLineData(6)} />
        <Card>
          <Card.Title
            title="Samples Timeline"
            subtitle={subtitle}
            right={() => <IconMenuButton size="small" menuItems={menuItems} />}
          />
          <Card.Content>
            {isLoading ? (
              <View style={{ padding: 20 }}>
                <ActivityIndicator
                  animating={true}
                  color={theme.colors.primary}
                  style={{ paddingVertical: 10 }}
                />
              </View>
            ) : (
              <Tabs defaultActiveIndex={1}>
                <Tabs.ButtonsContainer>
                  <Tabs.Button text="All" />
                  <Tabs.Button text="Transactions" />
                  <Tabs.Button text="Orders" />
                </Tabs.ButtonsContainer>
                <Tabs.Container>
                  <Tabs.Item>
                    <TimelineComponent
                      items={allItemsList}
                      navigation={this.props.navigation}
                    />
                  </Tabs.Item>
                  <Tabs.Item>
                    <TimelineComponent
                      items={transactionsList}
                      navigation={this.props.navigation}
                    />
                    {transactionsListId && transactionsList.length ? (
                      <ViewAll
                        url={`{EndPoint}&retURL=%2Flightning%2Fo%2F${NAMESPACE}SampleTransaction__c%2Flist%3FfilterName%3D${transactionsListId}`}
                      />
                    ) : null}
                  </Tabs.Item>
                  <Tabs.Item>
                    <TimelineComponent
                      items={ordersList}
                      navigation={this.props.navigation}
                    />
                    {ordersListId && ordersList.length ? (
                      <ViewAll
                        url={`{EndPoint}&retURL=%2Flightning%2Fo%2F${NAMESPACE}SampleOrder__c%2Flist%3FfilterName%3D${ordersListId}`}
                      />
                    ) : null}
                  </Tabs.Item>
                </Tabs.Container>
              </Tabs>
            )}
          </Card.Content>
        </Card>
      </>
    );
  }
}

export default () => {
  const navigation = useNavigation();
  const route = useRoute();
  const theme = useTheme();

  return <SamplesTimelineWidget navigation={navigation} route={route} theme={theme} />;
};
