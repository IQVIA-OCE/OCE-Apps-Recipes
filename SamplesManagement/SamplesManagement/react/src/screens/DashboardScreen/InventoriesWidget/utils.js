import moment from 'moment';
import { INVENTORY_STATUS } from '../../InventoryScreen/constants';
import { INVENTORY_FORM_TYPE } from '../../../constants/Inventories';
import { INVENTORY_TYPE } from './constants';
import { NAMESPACE } from '../../../constants/constants';

export const normalizeInventories = recordTypesData => inventoriesData => {
  let data = {
    all: [],
  };
  if (
    recordTypesData &&
    recordTypesData.length &&
    inventoriesData &&
    inventoriesData.length
  ) {
    data = inventoriesData.reduce((acc, item) => {
      const type = recordTypesData.find(type => item.RecordTypeId === type.Id);
      if (!acc[type.DeveloperName]) {
        acc[type.DeveloperName] = [];
      }
      item.DeveloperName = type.DeveloperName;
      acc[type.DeveloperName].push(item);
      return acc;
    }, {});
    data.all = inventoriesData;
  }

  return data;
};

export const normalizeRecordTypes = recordTypesData => {
  let data = {
    all: [],
  };
  if (recordTypesData && recordTypesData.length) {
    data = recordTypesData.reduce((acc, item) => {
      acc[item.DeveloperName] = item;
      return acc;
    }, {});
    data.all = recordTypesData;
  }

  return data;
};

export const getMenuItems = (navigation, { inventories, recordTypes }) => {
  if (
    inventories &&
    inventories.InitialInventory &&
    inventories.InitialInventory.length
  ) {
    const item = inventories.InitialInventory.find(
      el => el[`${NAMESPACE}Status__c`] !== INVENTORY_STATUS.submitted
    );
    if (item) {
      return [
        {
          text: INVENTORY_TYPE[item.DeveloperName],
          onPress: () =>
            navigation.navigate('Inventory', {
              id: item.Id,
              type:
                item[`${NAMESPACE}Status__c`] === INVENTORY_STATUS.inProgress
                  ? INVENTORY_FORM_TYPE.edit
                  : INVENTORY_FORM_TYPE.editSaved,
            }),
        },
      ];
    } else {
      return [
        {
          text: INVENTORY_TYPE.AdHocInventory,
          onPress: () =>
            navigation.navigate('Inventory', {
              recordType: recordTypes.AdHocInventory,
            }),
        },
        {
          text: INVENTORY_TYPE.AuditedInventory,
          onPress: () =>
            navigation.navigate('Inventory', {
              recordType: recordTypes.AuditedInventory,
            }),
        },
        {
          text: INVENTORY_TYPE.PeriodicInventory,
          onPress: () =>
            navigation.navigate('Inventory', {
              recordType: recordTypes.PeriodicInventory,
            }),
        },
      ];
    }
  }
  return [
    {
      text: INVENTORY_TYPE.InitialInventory,
      onPress: () =>
        navigation.navigate('Inventory', {
          recordType: recordTypes.InitialInventory,
        }),
    },
  ];
};

const DATE_FORMAT = 'MMM D, YYYY hh:mm A';

export const getDate = d => {
  const date = moment.utc(d);
  if (date.isValid()) {
    return date.format(DATE_FORMAT);
  } else {
    return moment.utc(d, 'DD-MM-YYYY hh:mm').format(DATE_FORMAT);
  }
};
