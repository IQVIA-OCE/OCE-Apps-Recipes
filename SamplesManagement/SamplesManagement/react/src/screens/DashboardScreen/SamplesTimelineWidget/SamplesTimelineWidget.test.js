import React from 'react';
import SamplesTimelineWidget from './SamplesTimelineWidget';
import { environment } from 'oce-apps-bridges';
import {
  fetchOrdersData,
  fetchTransactionsData,
} from '../../../api/SamplesTimeline';
import { mapOrders, mapTransactions } from './utils';
import moment from 'moment';
import { render } from '@testing-library/react-native';

jest.mock('moment', () => () => ({format: () => 'May 3, 2020 06:19 pm'}));
jest.mock('./utils');
jest.mock('../../../api/SamplesTimeline');

const navigation = {
  addListener: jest.fn().mockImplementation((_, fn) => ({ remove: jest.fn() })),
};

describe('SamplesTimelineWidget', () => {
  beforeEach(() => {
    environment.userID = jest.fn().mockReturnValue('1');
    fetchTransactionsData.mockReturnValue([
      {
        accountId: '1',
        accountName: 'name',
        address: null,
        conditionOfPackage: 'Undamaged',
        detailsCount: 1,
        fromSalesRepId: '0056F00000A4qfMQAR',
        fromSalesRepName: 'Name',
        id: 'a510w000000CgZQAA0',
        lastModifiedDate: '2020-05-20T11:55:45.0000000',
        receivedDate: '2020-05-20',
        recordTypeDevName: 'DeveloperName',
        recordTypeId: undefined,
        recordTypeName: 'Name',
        shipmentDate: null,
        status: 'In Progress',
        toSalesRepId: '0056F00000B45ezQAB',
        toSalesRepName: 'Name',
        transactionDateTime: '2020-05-20T10:45:12.0000000',
        transactionRepId: null,
        transactionRepName: 'Name',
      },
    ]);
    fetchOrdersData.mockReturnValue([
      {
        detailsCount: 1,
        id: '1',
        isUrgent: 'urgent',
        lastModifiedDate: undefined,
        recordTypeDevName: 'Order',
        recordTypeName: 'Order',
        status: 'status',
      },
    ]);
    mapTransactions.mockImplementation(d => d);
    mapOrders.mockImplementation(d => d);
  });

  it('Should render SamplesTimelineWidget component', async () => {
    const { getByText } = render(
      <SamplesTimelineWidget navigation={navigation} />
    );

    expect(getByText(/Samples Timeline/));
  });
});
