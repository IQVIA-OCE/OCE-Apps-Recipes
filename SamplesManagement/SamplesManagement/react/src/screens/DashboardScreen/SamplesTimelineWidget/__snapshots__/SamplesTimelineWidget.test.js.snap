// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SamplesTimelineWidget Should render SamplesTimelineWidget component 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "borderRadius": 4,
      "elevation": 2,
      "shadowColor": "#000000",
      "shadowOffset": Object {
        "height": 0.75,
        "width": 0,
      },
      "shadowOpacity": 0.24,
      "shadowRadius": 1.5,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingLeft": 16,
          },
          Object {
            "height": 72,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flex": 1,
              "flexDirection": "column",
              "height": 40,
              "justifyContent": "space-around",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                  "textAlign": "left",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "minHeight": 20,
                    },
                    Object {
                      "marginBottom": 2,
                    },
                    undefined,
                  ],
                ],
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "minHeight": 20,
                    },
                    Object {
                      "marginBottom": 2,
                    },
                    undefined,
                  ],
                ],
              ],
            ]
          }
        >
          Samples Timeline
        </Text>
      </View>
      <View>
        <View
          collapsable={false}
        >
          <View
            accessibilityComponentType="button"
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessibilityTraits="button"
            accessible={true}
            centered={true}
            focusable={true}
            hitSlop={
              Object {
                "bottom": 6,
                "left": 6,
                "right": 6,
                "top": 6,
              }
            }
            onClick={[Function]}
            onLayout={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            pointerEvents="box-only"
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                      "margin": 6,
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 16.5,
                      "height": 33,
                      "width": 33,
                    },
                    undefined,
                    undefined,
                  ],
                ]
              }
            >
              <View>
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  style={
                    Array [
                      Object {
                        "color": "#595959",
                        "fontSize": 22,
                      },
                      Array [
                        Object {
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  ï‡™
                </Text>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "bottom": 0,
                    "left": 0,
                    "overflow": "hidden",
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "borderRadius": 0,
                  },
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "paddingHorizontal": 16,
          },
          Object {
            "paddingVertical": 16,
          },
          undefined,
        ]
      }
    >
      <View>
        <RCTScrollView
          alwaysBounceHorizontal={true}
          horizontal={true}
          onLayout={[Function]}
          showsHorizontalScrollIndicator={false}
          style={
            Array [
              Object {
                "borderBottomWidth": 1,
                "borderColor": "rgba(0, 0, 0, 0.2)",
                "flexDirection": "row",
                "marginBottom": 20,
                "maxWidth": "100%",
                "paddingBottom": 10,
              },
              undefined,
            ]
          }
        >
          <View>
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onLayout={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "marginRight": 25,
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "rgb(0, 0, 0)",
                      "fontSize": 14,
                      "fontWeight": "700",
                    },
                    undefined,
                  ]
                }
              >
                All
              </Text>
              <View
                style={
                  Object {
                    "backgroundColor": "#0768FD",
                    "borderRadius": 4,
                    "bottom": -13,
                    "height": 5,
                    "left": 0,
                    "position": "absolute",
                    "width": "100%",
                  }
                }
              />
            </View>
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onLayout={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "marginRight": 25,
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "rgba(0, 0, 0, 0.7)",
                      "fontSize": 14,
                      "fontWeight": "600",
                    },
                    undefined,
                  ]
                }
              >
                Transactions
              </Text>
            </View>
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onLayout={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "marginRight": 0,
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "rgba(0, 0, 0, 0.7)",
                      "fontSize": 14,
                      "fontWeight": "600",
                    },
                    undefined,
                  ]
                }
              >
                Orders
              </Text>
            </View>
          </View>
        </RCTScrollView>
        <View>
          <View>
            <View />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
